File: 剑指Offer/字符串的排列.py
class Solution:
    def permutation(self, s: str) -> List[str]:
        c, res = list(s), []

        def dfs(x):
            if x == len(c) - 1:
                # 已经完成了一组排列
                res.append(''.join(c))
                return
            
            seen = set() # 用于去重
            for i in range(x, len(c)):
                if c[i] in seen:
                    continue
                seen.add(c[i])
                c[x], c[i] = c[i], c[x]
                dfs(x + 1)
                c[x], c[i] = c[i], c[x]
        
        dfs(0)
        return res

File: 剑指Offer/字符串的排列.cpp
class Solution {
    vector<string> res;
    void dfs(string& s, int index) {
        if (index == s.length() - 1) {
            res.emplace_back(s);
            return;
        }
        set<char> seen;
        for (int i = index + 1; i < s.length(); i++) {
            if (seen.find(s[i]) != seen.end()){
                continue; // 重复，减枝
            }
            seen.insert(s[i]);
            swap(s[index], s[i]);
            this->dfs(s, index + 1);
            swap(s[index], s[i]);
        }
    }

public:
    vector<string> permutation(string s) {
        this->dfs(s, 0);
        return res;
    }
};